DynWrite implementation

EnderGnar opened the pull request:
DynWrite is like List a type that can be used in specr and gets garbage collected.
It can be used to safely use something implementing `trait Write` for rust within specr.

EnderGnar committed:
DynWrite added.

EnderGnar committed:
WriteWrapper hash based on pointer.

memoryleak47 commented:
I think there is one problem with this solution.
If the object given into `DynWrite::new` contains for example a `GcCow`, this internal object will be freed prematurely when the GC runs.

Generally, the points_to implementation should always call `points_to` for all fields inside of a struct to collect all recursively contained GcCows. In this case, this is missing for `WriteWrapper`.

If this is intentional, and giving a GC'ed object into `DynWrite::new` is considered a misuse of the API, (this should probably be a comment)
then this PR could be somewhat simplified by letting `DynWrite` store the ref-counted `dyn Write` directly.

I personally think `trait GcWrite: GcCompat + Write` is the cleaner solution, as it allows to store GC'ed objects within `DynWriter`.
Another, even simpler alternative might be to have an `enum` instead of the dyn trait,
as there probably are a handful of `DynWrite` usages anyways.

RalfJung commented:
I'd prefer `dyn` over an `enum`.

But I don't understand why there are two separate types being defined here, spread across 3 files.

EnderGnar committed:
Remove of Wrapper & Intro of GcWrite

EnderGnar committed:
specr-transpile no_auto_derive

EnderGnar commented:
I now reworked `DynWrite` to work with a single new class. Now it also requires `GcWrite`.

specr-transpile supports `#[no_auto_derive]` for types that don't need them. This is only needed for `Machine` for now. But since it still needs `Clone`, `Copy`, and especially `GcCompat`. I think this is not good since GcCompat should be hidden. Maybe we split the Auto Encode attributes into multiple lists.

RalfJung commented:
> specr-transpile supports #[no_auto_derive] for types that don't need them. This is only needed for Machine for now. But since it still needs Clone, Copy, and especially GcCompat. I think this is not good since GcCompat should be hidden. Maybe we split the auto derive traits into multiple lists.

I'd suggest to rename the attribute to `no_obj`, and make it still derive Clone, Copy, and GcCompat, but not the other traits.

RalfJung commented on libspecr/src/write/mod.rs:
```diff
@@ -0,0 +1,50 @@
+pub trait GcWrite: GcCompat + Write {}
+impl<T> GcWrite for T where T: GcCompat + Write {}
+
+#[derive(Copy, Clone, GcCompat)]
+/// Garbage-collected datastructure representing a write stream and implementing `Copy`.
```
```suggestion
/// Garbage-collected data structure representing a write stream and implementing `Copy`.
#[derive(Copy, Clone, GcCompat)]
```
Doc comments usually go before attributes

EnderGnar committed:
`mark_and_sweep` with reference.

EnderGnar committed:
Doc comments

Co-authored-by: Ralf Jung <post@ralfj.de>

RalfJung commented:
Die libspecr changes look good. For specr-transpile, does this now requires any `derive`s in MiniRust?

EnderGnar committed:
no_obj

EnderGnar committed:
Merge branch 'write-stream' of https://github.com/EnderGnar/minirust-tooling into write-stream

EnderGnar commented:
I have now changed `no_auto_derive` to `no_obj`. In general `Clone`, `GcCompat`, and `Debug` are derived. All other traits are only derived for objects. 
This means that we once again don't need derives in MiniRust.

RalfJung commented:
Where do we still need `Clone`? Having `Clone` but not `Copy` doesn't seem to make a ton of sense

EnderGnar commented:
We currently don't need `Clone` for the machine.

EnderGnar committed:
Clone removed from general traits.
